spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -488,123 +563,155 @@ NOTE: It is up to individual implementations how they will provide access-contro
 
     // call modules onChanUpgradeAck callback
+    // module can error on counterparty version
+    // ACK should not change state to the new parameters yet
+    // as that will happen on the onChanUpgradeOpen callback
     module = lookupModule(portIdentifier)
     err = module.onChanUpgradeAck(
         portIdentifier,
         channelIdentifier,
-        counterpartyChannel.channelIdentifier,
-        counterpartyChannel.version
+        counterpartyUpgrade.version
     )
     // restore channel if callback returned error
     if err != nil {
         restoreChannel(portIdentifier, channelIdentifier)
         return
     }
 
-    // upgrade is complete
-    // set channel to OPEN and remove unnecessary state
-    currentChannel.state = OPEN
-    provableStore.set(channelPath(portIdentifier, channelIdentifier), currentChannel)
-    provableStore.delete(channelUpgradeTimeoutPath(portIdentifier, channelIdentifier))
-    privateStore.delete(channelRestorePath(portIdentifier, channelIdentifier))
+    // if no error, agree on final version
+    upgrade.version = counterpartyUpgrade.version
+    provableStore.set(channelUpgradePath(portIdentifier, channelIdentifier), upgrade)
+
+    // refresh channel
+    currentChannel = provableStore.get(channelPath(portIdentifier, channelIdentifier))
+
+    // if both sides have already flushed then open the upgrade handshake immediately
+    if  currentChannel.state == FLUSHCOMPLETE && counterpartyFlushStatus == FLUSHCOMPLETE {
+        openChannelHandshake(portIdentifier, channelIdentifier)
+        module.onChanUpgradeOpen(portIdentifier, channelIdentifier)
+    }
 }
 ```
 
 ```typescript
-function chanUpgradeConfirm(
+function chanUpgradeOpen(
     portIdentifier: Identifier,
     channelIdentifier: Identifier,
-    counterpartyChannel: ChannelEnd,
+    counterpartyChannelState: ChannelState,
     proofChannel: CommitmentProof,
-    proofUpgradeError: CommitmentProof,
-    proofUpgradeSequence: CommitmentProof,
     proofHeight: Height,
 ) {
-    // current channel is in TRYUPGRADE
-    currentChannel = provableStore.get(channelPath(portIdentifier, channelIdentifier))
-    abortTransactionUnless(channel.state == TRYUPGRADE)
+    // if packet commitments are not empty then abort the transaction
+    abortTransactionUnless(pendingInflightPackets(portIdentifier, channelIdentifier))
 
-    // counterparty must be in OPEN state
-    abortTransactionUnless(counterpartyChannel.State == OPEN)
-
-    // get current sequence
-    sequence = provableStore.get(channelUpgradeSequencePath(portIdentifier, channelIdentifier))
+    // currentChannel must be in TRYUPGRADE or ACKUPGRADE and have completed flushing
+    currentChannel = provableStore.get(channelPath(portIdentifier, channelIdentifier))
+    abortTransactionUnless(currentChannel.state == TRYUPGRADE || currentChannel.state == ACKUPGRADE)
+    abortTransactionUnless(currentChannel.flushStatus == FLUSHCOMPLETE)
```

Comment:
is there another way to set our `currentChannel.flushStatus` to `FLUSHCOMPLETE` besides this code snippet in `StartFlushingUpgradeHandshake`? 

```
    // if there are no in-flight packets on our end, we can automatically go to FLUSHCOMPLETE
    if pendingInflightPackets(portIdentifier, channelIdentifier) == nil {
        currentChannel.flushState = FLUSHCOMPLETE
    }
   ```

we call this in `Ack` and `Try` but we can't recall either of them once our channel state is in `UPGRADEACK`.

Reply:
This will be done in the packet handlers in a future spec PR. this pr is purely focused on upgrade handshake changes to scope the changes.

Reply:
is it important in this spec that these handlers are at least mentioned? otherwise the logic for the onUpgradeOpen seems imcomplete.